lib/utils/signature.js

```diff
@@ -0,0 +1,395 @@
+const common = require('../protocol/common');
+
+/**
+ * Size of the curve
+ * @const {Number}
+ */
+
+const SIZE = 32;
+
+/**
+ * 32 byte number zero
+ * @const {Buffer}
+ */
+
+const ZERO = Buffer.alloc(32, 0x00);
+
+/**
+ * This based on bcrypto/lib/internal/signature
+ * Adds recid on top.
+ * @property {Buffer} r
+ * @property {Buffer} s
+ * @property {Number} recid
+ */
+
+class LedgerSignature {
+  /**
+   * Create LedgerSignature object.
+   *
+   * @param {Buffer?} r
+   * @param {Buffer?} s
+   * @param {Number?} recid
+   */
+
+  constructor(r, s, recid) {
+    this.r = ZERO;
+    this.s = ZERO;
+    this.recid = -1;
+
+    if (r)
+      this.setR(r);
+
+    if (s)
+      this.setS(s);
+
+    if (recid)
+      this.setRecid(recid);
+  }
+
+  /**
+   * Inject options
+   * @param {Object} options
+   * @param {Buffer?} options.r
+   * @param {Buffer?} options.s
+   * @param {Number?} options.recid
+   * @returns {LedgerSignature}
+   */
+
+  fromOptions(options) {
+    assert(typeof options === 'object');
+
+    if (options.r)
+      this.setR(options.r);
+
+    if (options.s)
+      this.setS(options.s);
+
+    if (options.recid)
+      this.setRecid(options.recid);
+
+    return this;
+  }
+
+  /**
+   * Set R.
+   * @param {Buffer} r
+   * @returns {LedgerSignature}
+   */
+
+  setR(r) {
+    assert(Buffer.isBuffer(r));
+    assert(r.length === SIZE);
+
+    this.r = r;
+
+    return this;
+  }
+
+  /**
+   * Set S.
+   * @param {Buffer} s
+   * @returns {LedgerSignature}
+   */
+
+  setS(s) {
+    assert(Buffer.isBuffer(s));
+    assert(s.length === SIZE);
+
+    this.s = s;
+
+    return this;
+  }
+
+  /**
+   * Set Recovery ID.
+   * @param {Number} recid
+   * @returns {LedgerSignature}
+   */
+
+  setRecid(recid) {
+    assert(typeof recid === 'number', 'recid must be a number.');
+    assert(recid >= 0 && recid <= 3,
+      'recid must be a number between 0 and 3.');
+
+    this.recid = recid;
+
+    return this;
+  }
+
+  /**
+   * Check equality with other signature
+   * @param {LedgerSignature} sig
+   * @returns {Boolean}
+   */
+
+  equals(sig) {
+    assert(sig instanceof LedgerSignature);
+
+    return this.recid === sig.recid
+      && this.s.equals(sig.s)
+      && this.r.equals(sig.r);
+  }
+
+  /**
+   * Encode to RS buffer.
+   * @returns {Buffer} - R || S (32 + 32 bytes)
+   */
+
+  encode() {
+    const raw = Buffer.allocUnsafe(SIZE * 2);
+
+    this.r.copy(raw, 0);
+    this.s.copy(raw, SIZE);
+
+    return raw;
+  }
+
+  /**
+   * Decode RS buffer.
+   * @param {Buffer} data - R || SS (32 + 32 bytes)
+   * @returns {LedgerSignature}
+   */
+
+  decode(data) {
+    assert(Buffer.isBuffer(data), 'sig must be a buffer.');
+    assert(data.length === SIZE * 2);
+
+    this.r = data.slice(0, SIZE);
+    this.s = data.slice(SIZE, SIZE * 2);
+
+    return this;
+  }
+
+  /**
+   * Deserialize DER encoded buffer to LedgerSignature.
+   * @param {Buffer} der
+   * @returns {LedgerSignature}
+   */
+
+  fromDER(der) {
```

Comment:
`fromDER` will not set the recovery param and since it is initialized to `-1`, it will cause `recover` to always return `null`. It is possible to brute force the recovery id since there are only 4 of them, [0,3]